src/Npgsql/Internal/TypeHandlers/MultirangeHandler.cs

```diff
@@ -0,0 +1,121 @@
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Npgsql.BackendMessages;
+using Npgsql.Internal.TypeHandling;
+using Npgsql.PostgresTypes;
+using NpgsqlTypes;
+
+namespace Npgsql.Internal.TypeHandlers
+{
+    /// <summary>
+    /// An interface implementing by <see cref="MultirangeHandler{TElement}"/>, exposing the handler's supported range
+    /// CLR types.
+    /// </summary>
+    public interface IMultirangeHandler
+    {
+        /// <summary>
+        /// Exposes the multirange CLR types supported by this handler.
+        /// </summary>
+        Type[] SupportedMultirangeClrTypes { get; }
+    }
+
+    public partial class MultirangeHandler<TElement> : NpgsqlTypeHandler<NpgsqlRange<TElement>[]>, 
+        INpgsqlTypeHandler<List<NpgsqlRange<TElement>>>, IMultirangeHandler
+    {
+        /// <summary>
+        /// The type handler for the range that this multirange type holds
+        /// </summary>
+        readonly RangeHandler<TElement> _rangeHandler;
+
+        /// <inheritdoc />
+        public Type[] SupportedMultirangeClrTypes { get; }
+
+        /// <inheritdoc />
+        public MultirangeHandler(PostgresMultirangeType pgMultirangeType, RangeHandler<TElement> rangeHandler)
+            : this(pgMultirangeType, rangeHandler, new[] { typeof(NpgsqlRange<TElement>[]), typeof(List<NpgsqlRange<TElement>>) }) {}
+
+        /// <inheritdoc />
+        protected MultirangeHandler(
+            PostgresMultirangeType pgMultirangeType, RangeHandler<TElement> rangeHandler, Type[] supportedSubtypeClrTypes)
+        {
+            PostgresType = pgMultirangeType;
+            _rangeHandler = rangeHandler;
+            SupportedMultirangeClrTypes = supportedSubtypeClrTypes;
+        }
+
+        public override async ValueTask<NpgsqlRange<TElement>[]> Read(
+            NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription = null)
+        {
+            await buf.Ensure(4, async);
+
+            var numRanges = buf.ReadInt32();
+            var multirange = new NpgsqlRange<TElement>[numRanges];
+
+            for (var i = 0; i < numRanges; i++)
+            {
+                var rangeLen = buf.ReadInt32();
```

Comment:
Shouldn't `Ensure(4)` be called at the start of every loop?

Reply:
Good batch...